package/samplers/carbo/sampler.py

```diff
@@ -0,0 +1,280 @@
+    clipped_values = convert_inf(values)
+    means = np.mean(clipped_values, axis=0)
+    stds = np.std(clipped_values, axis=0)
+    standardized_values = (clipped_values - means) / np.maximum(EPS, stds)
+    return standardized_values, means, stds
+
+
+def unnormalize_params(
+    params: np.ndarray, is_log: np.ndarray, lows: np.ndarray, highs: np.ndarray
+) -> np.ndarray:
+    assert len(params.shape) == 2
+    results = params * (highs - lows) + lows
+    log_lows = np.log(lows[is_log])
+    log_highs = np.log(highs[is_log])
+    results[:, is_log] = np.exp(params[:, is_log] * (log_highs - log_lows) + log_lows)
+    return results
+
+
+def normalize_params(
+    params: np.ndarray, is_log: np.ndarray, lows: np.ndarray, highs: np.ndarray
+) -> np.ndarray:
+    assert len(params.shape) == 2
+    results = (params - lows) / (highs - lows)
+    log_lows = np.log(lows[is_log])
+    log_highs = np.log(highs[is_log])
+    results[:, is_log] = (np.log(params[:, is_log]) - log_lows) / (log_highs - log_lows)
+    return results
+
+
+def default_log_prior(kernel_params: KernelParamsTensor) -> torch.Tensor:
+    def gamma_log_prior(x: torch.Tensor, concentration: float, rate: float) -> torch.Tensor:
+        return (concentration - 1) * torch.log(x) - rate * x
+
+    return (
+        -(
+            0.1 / kernel_params.inverse_squared_lengthscales
+            + 0.1 * kernel_params.inverse_squared_lengthscales
+        ).sum()
+        + gamma_log_prior(kernel_params.kernel_scale, 2, 1)
+        + gamma_log_prior(kernel_params.noise_var, 1.1, 30)
+    )
+
+
+def _get_dist_info_as_arrays(
+    search_space: dict[str, BaseDistribution],
+) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
+    dim = len(search_space)
+    is_log = np.zeros(dim, dtype=bool)
+    lows = np.empty(dim, dtype=float)
+    highs = np.empty(dim, dtype=float)
+    for d, dist in enumerate(search_space.values()):
+        assert isinstance(dist, optuna.distributions.FloatDistribution)
+        is_log[d] = dist.log
+        lows[d] = dist.low
+        highs[d] = dist.high
+    return lows, highs, is_log
+
+
+def _get_params_array(
+    trials: list[FrozenTrial], search_space: dict[str, BaseDistribution]
+) -> np.ndarray:
+    params = np.empty((len(trials), len(search_space)), dtype=float)
+    for i, t in enumerate(trials):
+        for d, (name, dist) in enumerate(search_space.items()):
+            params[i, d] = t.params[name]
+    return params
+
+
+class CARBOSampler(BaseSampler):
+    def __init__(
+        self,
+        *,
+        seed: int | None = None,
+        independent_sampler: BaseSampler | None = None,
+        n_startup_trials: int = 10,
+        deterministic_objective: bool = False,
+        constraints_func: Callable[[FrozenTrial], Sequence[float]] | None = None,
+        rho: float = 1e3,
+        beta: float = 4.0,
+        local_ratio: float = 0.1,
+        n_local_search: int = 10,
+    ) -> None:
+        self._rng = np.random.RandomState(seed)
```

Comment:
Although this is optional, `LazyRandomState` will be required for proper parallelization.

Reply:
Applied it to pass the unit tests!